File: flipt-client-ruby/spec/evaluation_spec.rb
# frozen_string_literal: true

require_relative '../lib/flipt_client'

RSpec.describe Flipt::EvaluationClient do
  before(:all) do
    url = ENV.fetch('FLIPT_URL', 'http://localhost:8080')
    auth_token = ENV.fetch('FLIPT_AUTH_TOKEN', 'secret')
    @client = Flipt::EvaluationClient.new('default', { url: url, auth_token: auth_token })
  end

  describe '#evaluate_variant' do
    it 'returns a variant result' do
      resp = @client.evaluate_variant({ flag_key: 'flag1', entity_id: 'someentity', context: { "fizz": 'buzz' } })

      expect(resp).to_not be_nil
      expect(resp['status']).to eq('success')
      expect(resp['error_message']).to be_nil
      expect(resp['result']['flag_key']).to eq('flag1')
      expect(resp['result']['match']).to eq(true)
      expect(resp['result']['reason']).to eq('MATCH_EVALUATION_REASON')
      expect(resp['result']['variant_key']).to eq('variant1')
      expect(resp['result']['segment_keys']).to eq(['segment1'])
    end
  end

  describe '#evaluate_boolean' do
    it 'returns a boolean result' do
      resp = @client.evaluate_boolean({ flag_key: 'flag_boolean', entity_id: 'someentity', context: { "fizz": 'buzz' } })

      expect(resp).to_not be_nil
      expect(resp['status']).to eq('success')
      expect(resp['error_message']).to be_nil
      expect(resp['result']['flag_key']).to eq('flag_boolean')
      expect(resp['result']['enabled']).to eq(true)
      expect(resp['result']['reason']).to eq('MATCH_EVALUATION_REASON')
    end
  end

  describe '#evaluate_variant failure' do
    it 'gracefully handles failures for variant flag evaluation' do
      resp = @client.evaluate_variant({ flag_key: 'nonexistent', entity_id: 'someentity', context: { "fizz": 'buzz' } })

      expect(resp).to_not be_nil
      expect(resp['result']).to be_nil
      expect(resp['status']).to eq('failure')
      expect(resp['error_message']).to eq('invalid request: failed to get flag information default/nonexistent')
    end
  end

  describe '#evaluate_boolean failure' do
    it 'gracefully handles failures for boolean flag evaluation' do
      resp = @client.evaluate_boolean({ flag_key: 'nonexistent', entity_id: 'someentity', context: { "fizz": 'buzz' } })

      expect(resp).to_not be_nil
      expect(resp['result']).to be_nil
      expect(resp['status']).to eq('failure')
      expect(resp['error_message']).to eq('invalid request: failed to get flag information default/nonexistent')
    end
  end
end

File: flipt-engine-wasm/integration/spec/evaluation.spec.js
import data from "./support/state.json" assert { type: "json" };
import { Engine } from "../../pkg/flipt_engine_wasm.js";

describe("Evaluation", () => {
  let engine;

  beforeEach(() => {
    engine = new Engine("default", data);
  });

  it("should evaluate a boolean flag", async () => {
    const response = engine.evaluate_boolean({
      flag_key: "flag_boolean",
      entity_id: "user_1",
      context: {},
    });

    expect(response).not.toBeNull();
    expect(response.status).toEqual("success");
    expect(response.result).not.toBeNull();

    const result = response.result;
    expect(result.enabled).toBe(true);
    expect(result.flag_key).toEqual("flag_boolean");
    expect(result.reason).toEqual("MATCH_EVALUATION_REASON");
  });

  it("should evaluate a multivariate flag", async () => {
    const response = engine.evaluate_variant({
      flag_key: "flag1",
      entity_id: "user_1",
      context: {
        fizz: "buzz",
      },
    });

    expect(response).not.toBeNull();
    expect(response.status).toEqual("success");
    expect(response.result).not.toBeNull();

    const result = response.result;
    expect(result.match).toBe(true);
    expect(result.flag_key).toEqual("flag1");
    expect(result.reason).toEqual("MATCH_EVALUATION_REASON");
    expect(result.variant_key).toEqual("variant1");
    expect(result.segment_keys).toEqual(["segment1"]);
  });
});
